labrad/backend.py

```diff
@@ -368,56 +374,55 @@ def raiseException(self):
-        if not hasattr(error, 'raiseException'):
-            error = Failure(error)
-        self.result = error
-        self.ready.put(self)
+class Future(concurrent.futures.Future):
+    """Backwards compatible subclass of concurrent.futures.Future.
+
+    The standard concurrent.futures.Future class uses .result() to get the
+    final result, whereas the old labrad.backend.Future class used .wait().
+    """
+
+    @staticmethod
+    def then(prev, f, *args, **kw):
+        """Return a future that applies a function to another future.
+
+        Args:
+            prev (concurrent.futures.Future): The future whose result we wish
+                to modify.
+            f (callable): Function to apply to the result of prev. This
+                function will be called like f(result, *args, **kw) where
+                result is the value produced by prev.
+            args: Positional args to be passed to f.
+            kw: Keyword args to be passed to f.
+
+        Returns:
+            A new Future whose result will be the given function applied to the
+            result of prev. If prev is cancelled or fails with an exception, or
+            if the function f raises an exception, then the returned future
+            will similarly be cancelled.
+        """
+        future = Future()
+        def handle_result(prev):
+            if not future.set_running_or_notify_cancel():
+                return
+            if prev.cancelled():
+                future.set_exception(concurrent.futures.CancelledError())
+                return
+            error = prev.exception()
+            if error is not None:
+                future.set_exception(error)
+                return
+            try:
+                result = f(prev.result(), *args, **kw)
+                future.set_result(result)
+            except Exception as e:
+                future.set_exception(e)
+        prev.add_done_callback(handle_result)
+        return future
 
     def wait(self):
-        if self.done:
-            return self.result
-        while True:
-            f = self.ready.get()
-            f.done = True
-            result = f.result
-            if hasattr(result, 'raiseException'):
-                # this can be a Twisted Failure or our
-                # own Failure class, as defined above
-                # If any Future in the queue fails,
-                # we immediately bail.
-                result.raiseException()
-            else:
-                for func, args, kw in f.callbacks:
-                    result = func(result, *args, **kw)
-                f.result = result
-            if f is self:
-                return result
-
-    def __repr__(self):
-        if self.done:
-            return '<Future: result=%r>' % (self.result,)
-        else:
-            return '<Future: pending...>'
+        """Alias for the result() method; for backwards compatibility."""
+        warnings.warn(".wait() is deprecated; use .result() instead.")
```

Comment:
I had that initially, but the problem is that DeprecationWarning is filtered by default, so users probably won't see it.